File: app/src/main/java/codetagging/simflect/builder/TypeReflection.java
package codetagging.simflect.builder;

import android.content.Context;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;

/**
 * Created by govind on 23/05/16.
 */
public class TypeReflection {

    public static Object get(String className, String methodName, Object param, Class classDataType,  Context context)
            throws ClassNotFoundException, NoSuchMethodException, Exception {
        Object returnObject = null;
        try {
            Class classType = Class.forName(className);
            Object classObject = null;
            Constructor[] constructors = classType.getConstructors();
            for (Constructor constructor: constructors
                    ) {
                Class[] parameterTypes = constructor.getParameterTypes();
                if(parameterTypes != null) {
                    if(parameterTypes.length == 1 && parameterTypes[0] == Context.class) {
                        classObject = constructor.newInstance(context);
                        break;
                    }
                }
            }
            Method method = classType.getDeclaredMethod(methodName, classDataType);
            returnObject = method.invoke(classObject, param);
        }
        catch (ClassNotFoundException classNotFoundException) {
            throw new ClassNotFoundException(className);
        }
        catch (NoSuchMethodException noSuchMethodException) {
            throw new NoSuchMethodException(methodName);
        }
        catch (Exception e) {
            throw new Exception(e.getMessage());
        }
        return returnObject;
    }

    public static Object getDefault(String className, String methodName, int instanceParam)
            throws ClassNotFoundException, NoSuchMethodException, Exception {
        Object returnObject = null;
        try {
            Class classType = Class.forName(className);
            Object classObject = null;
            classObject = classType.getDeclaredMethod("getDefault", int.class).invoke(null, instanceParam);
            /*Constructor[] constructors = classType.getConstructors();
            for (Constructor constructor: constructors
                    ) {
                Class[] parameterTypes = constructor.getParameterTypes();
                if(parameterTypes != null) {
                    if(parameterTypes.length == 1 && parameterTypes[0] == int.class) {
                        classObject = constructor.newInstance(instanceParam);
                        break;
                    }
                }
            }*/
            Method method = classType.getDeclaredMethod(methodName);
            returnObject = method.invoke(classObject);
        }
        catch (ClassNotFoundException classNotFoundException) {
            throw new ClassNotFoundException(className);
        }
        catch (NoSuchMethodException noSuchMethodException) {
            throw new NoSuchMethodException(methodName);
        }
        catch (Exception e) {
            throw new Exception(e.getMessage());
        }
        return returnObject;
    }

    public static Object getClassStaticInstance(String className,
                                                String staticMethodName,
                                                Object param,
                                                Class dataType) throws ClassNotFoundException, IllegalArgumentException, Exception {
        Object classObject = null;
        try {
            Class classReference = Class.forName(className);
            if(param == null) {
                classObject = classReference.getMethod(staticMethodName, new Class<?>[0])
                        .invoke(null, new Object[0]);
            } else {
                classObject = classReference.getMethod(staticMethodName, dataType)
                        .invoke(null, param);
            }
        }
        catch (ClassNotFoundException e) {
            throw new ClassNotFoundException(e.getMessage());
        }
        catch (IllegalArgumentException e) {
            throw  new IllegalArgumentException(e.getMessage());
        }
        catch (Exception e) {
            throw new Exception(e.getMessage());
        }
        return classObject;
    }

    public static Object getClassContextInstance(String className, Context context) throws
            ClassNotFoundException, IllegalArgumentException, Exception {
        Object classObject = null;
        try {
            Class classReference = Class.forName(className);
            Constructor[] constructors = classReference.getConstructors();
            for (Constructor constructor: constructors
                    ) {
                Class[] parameterTypes = constructor.getParameterTypes();
                if(parameterTypes != null) {
                    if(parameterTypes.length == 1 && parameterTypes[0] == Context.class) {
                        classObject = constructor.newInstance(context);
                        break;
                    }
                }
            }
        }
        catch (ClassNotFoundException e) {
            throw new ClassNotFoundException(e.getMessage());
        }
        catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(e.getMessage());
        }
        catch (Exception e) {
            throw new Exception(e.getMessage());
        }
        return classObject;
    }

    public static Object invokeMethod(Object classObject,
                                      String methodName,
                                      Object param,
                                      Class dataType) throws NoSuchMethodException, IllegalArgumentException, Exception {
        Object returnObject = null;
        try {
            Method method = null;
            if(param == null) {
                method = classObject.getClass().getDeclaredMethod(methodName);
                returnObject = method.invoke(classObject);
            } else {
                method = classObject.getClass().getMethod(methodName, dataType);
                returnObject = method.invoke(classObject, param);
            }
        }
        catch (NoSuchMethodException e) {
            throw new NoSuchMethodException(e.getMessage());
        }
        catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(e.getMessage());
        }
        catch (Exception e) {
            throw new Exception(e.getMessage());
        }
        return returnObject;
    }

    public static boolean isClassAvailable(String className) {
        try {
            Class classReference = Class.forName(className);
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            return false;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public static boolean isMethodAvailable(Object classObject, String methodName, Class paramType) {
        Method method = null;
        try {
            if(paramType != null) {
                method = classObject.getClass().getMethod(methodName, new Class[] {paramType});
            } else {
                method = classObject.getClass().getMethod(methodName);
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
            return false;
        }
        return method != null;
    }

}

File: library/src/main/java/com/codetagging/simflect/util/TypeReflection.kt
package com.codetagging.simflect.util

import android.content.Context
import java.lang.reflect.Method

object TypeReflection {

    @Throws(ClassNotFoundException::class, NoSuchMethodException::class, Exception::class)
    operator fun get(className: String, methodName: String, param: Any, classDataType: Class<*>, context: Context): Any {
        var returnObject: Any? = null
        try {
            val classType = Class.forName(className)
            var classObject: Any? = null
            val constructors = classType.constructors
            for (constructor in constructors) {
                val parameterTypes = constructor.parameterTypes
                if (parameterTypes != null) {
                    if (parameterTypes.size == 1 && parameterTypes[0] == Context::class.java) {
                        classObject = constructor.newInstance(context)
                        break
                    }
                }
            }
            val method = classType.getDeclaredMethod(methodName, classDataType)
            returnObject = method.invoke(classObject, param)
        } catch (classNotFoundException: ClassNotFoundException) {
            throw ClassNotFoundException(className)
        } catch (noSuchMethodException: NoSuchMethodException) {
            throw NoSuchMethodException(methodName)
        } catch (e: Exception) {
            throw Exception(e.message)
        }

        return returnObject
    }

    @Throws(ClassNotFoundException::class, NoSuchMethodException::class, Exception::class)
    fun getDefault(className: String, methodName: String, instanceParam: Int): Any {
        var returnObject: Any? = null
        try {
            val classType = Class.forName(className)
            var classObject: Any? = null
            classObject = classType.getDeclaredMethod("getDefault", Int::class.javaPrimitiveType).invoke(null, instanceParam)
            val method = classType.getDeclaredMethod(methodName)
            returnObject = method.invoke(classObject)
        } catch (classNotFoundException: ClassNotFoundException) {
            throw ClassNotFoundException(className)
        } catch (noSuchMethodException: NoSuchMethodException) {
            throw NoSuchMethodException(methodName)
        } catch (e: Exception) {
            throw Exception(e.message)
        }

        return returnObject
    }

    @Throws(ClassNotFoundException::class, IllegalArgumentException::class, Exception::class)
    fun getClassStaticInstance(className: String?,
                               staticMethodName: String?,
                               param: Any?,
                               dataType: Class<*>?): Any {
        var classObject: Any? = null
        try {
            val classReference = Class.forName(className)
            if (param == null) {
                classObject = classReference.getMethod(staticMethodName, *arrayOfNulls<Class<*>>(0))
                        .invoke(null, *arrayOfNulls<Any>(0))
            } else {
                classObject = classReference.getMethod(staticMethodName, dataType)
                        .invoke(null, param)
            }
        } catch (e: ClassNotFoundException) {
            throw ClassNotFoundException(e.message)
        } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException(e.message)
        } catch (e: Exception) {
            throw Exception(e.message)
        }

        return classObject
    }

    @Throws(ClassNotFoundException::class, IllegalArgumentException::class, Exception::class)
    fun getClassContextInstance(className: String, context: Context): Any {
        var classObject: Any? = null
        try {
            val classReference = Class.forName(className)
            val constructors = classReference.constructors
            for (constructor in constructors) {
                val parameterTypes = constructor.parameterTypes
                if (parameterTypes != null) {
                    if (parameterTypes.size == 1 && parameterTypes[0] == Context::class.java) {
                        classObject = constructor.newInstance(context)
                        break
                    }
                }
            }
        } catch (e: ClassNotFoundException) {
            throw ClassNotFoundException(e.message)
        } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException(e.message)
        } catch (e: Exception) {
            throw Exception(e.message)
        }

        return classObject!!
    }

    @Throws(NoSuchMethodException::class, IllegalArgumentException::class, Exception::class)
    fun invokeMethod(classObject: Any?,
                     methodName: String?,
                     param: Any?,
                     dataType: Class<*>?): Any {
        var returnObject: Any? = null
        try {
            var method: Method? = null
            if (param == null) {
                method = classObject?.javaClass?.getDeclaredMethod(methodName)
                returnObject = method!!.invoke(classObject)
            } else {
                method = classObject?.javaClass?.getMethod(methodName, dataType)
                returnObject = method!!.invoke(classObject, param)
            }
        } catch (e: NoSuchMethodException) {
            throw NoSuchMethodException(e.message)
        } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException(e.message)
        } catch (e: Exception) {
            throw Exception(e.message)
        }

        return returnObject
    }

    fun isClassAvailable(className: String?): Boolean {
        try {
            val classReference = Class.forName(className)
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
            return false
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }

        return true
    }

    fun isMethodAvailable(classObject: Any?, methodName: String?, paramType: Class<*>?): Boolean {
        var method: Method? = null
        try {
            if (paramType != null) {
                method = classObject?.javaClass?.getMethod(methodName, *arrayOf(paramType))
            } else {
                method = classObject?.javaClass?.getMethod(methodName)
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
            return false
        }

        return method != null
    }

}
